proposals/apple-mdm/Fleet-MDM-Plan.md

```diff
@@ -0,0 +1,218 @@
+
+###### 1. `fleetctl apple-mdm enrollments dep set-token setup --enrollment=<ENROLLMENT_ID>`
+
+- Checks the `ENROLLMENT_ID` is a DEP enrollment.
+- Generates `fleet-mdm-apple-dep.pem` and `fleet-mdm-apple-dep.key`:
+	- Stores `fleet-mdm-apple-dep.pem` as a file.
+	- Keeps `fleet-mdm-apple-dep.key` in `~/.fleet/config` under <ENROLLMENT_ID>.
+
+###### 2. Upload PEM to Apple
+
+User uploads `fleet-mdm-apple-dep.pem` to https://business.apple.com, and downloads a `*.p7m` file. Let's call it `fleet-mdm-apple-dep-auth-token-encrypted.p7m`.
+
+###### 3. Finalize DEP setup
+
+`fleetctl apple-mdm enrollments dep set-token finalize --enrollment=<ENROLLMENT_ID> --encrypted-auth-token=fleet-mdm-apple-dep-auth-token-encrypted.p7m`
+	
+- Checks that the `ENROLLMENT_ID` is a DEP enrollment.
+- Decrypts the provided `fleet-mdm-apple-dep-auth-token-encrypted.p7m` with the `fleet-mdm-apple-dep.key` that corresponds to `<ENROLLMENT_ID>` from `~/.fleet/config`.
+- Calls new Fleet API that:
+	- Apply DEP configuration using token (loaded from apple_dep_enrollmentsID==<ENROLLMENT_ID>) and get a "depProfileUUID".
+	- Creates `dep_names` entry with such token + depProfileUUID (as assigner_profile_uuid) (encryption of the token in MySQL will be a TODO.).
+	- Link such `dep_names` entry to the given `<ENROLLMENT_ID>`.
+- If all looks good, then it removes `fleet-mdm-apple-dep.key` from `~/.fleet/config`.
+
+#### Custom commands for DEP enrollments
+
+- List DEP devices: DEP devices of that enrollment and their status (using the DEP proxy API).
+	`fleetctl apple-mdm enrollments dep list <ENROLLMENT_ID>`
+- Sync DEP profiles: makes sure to set the enroll profile for new devices in a DEP enrollment. (Fleet would sync all DEP enrollments automatically every 5m.)
+	`fleetctl apple-mdm enrollments dep sync-profiles <ENROLLMENT_ID>`
+
+#### DEP syncer
+
+Fleet must start a "DEP syncer" routine for each entry in `apple_dep_enrollments` that has a non-NULL `dep_names` (period of 5m).
+
+### Hosts
+
+Initially, we want to account for hosts that are MDM enrolled but not in osquery.
+TODO(Lucas): We'll need to test whether our system and UI support a `hosts` entry w/ an empty `osquery_host_id`...
+
+### MDM commands
+
+These are one-off commands like "DeviceLock", "RestartDevice", etc.
+
+Fleetctl commands:
+- Send command:
+	`fleetctl apple-mdm commands send --target-hosts=1,2,3 --command=foo.plist`
+(fleetctl would be agnostic to contents of "foo.plist")
+- List commands (with their status):
+	`fleetctl apple-mdm commands list`
+
+### MDM profiles
+
+A profile is defined as:
+- ID
+- UUID
+- Name
+- Payload (raw XML plist)
+- Team (NULL for global)
+
+For MVP-dogfood, we will only have global profiles, and as such, they are applied to all MDM enrolled hosts.
+
+TODO(Lucas): Stuff to solve around this feature:
+So, we have a list of all the profiles to apply to devices. How does Fleet know which profiles are
```

Comment:
This is interesting. Definitely is important to be built for end user product. But I don't think we need it for dogfood. Can we use osquery to confirm the settings are appropriately applied on this? That's actually a piece of dogfooding/pain that would be valuable learning for us.

Reply:
More context here:

One common complaint I hear during user interviews about existing MDMs is: "once I've sent the profiles, how do I know they've been installed? I get no feedback." One of the theses for doing MDM is the power of combining high visibility osquery w/ MDM in order to generate that feedback loop.

Reply:
We'll be defining the plan for this soon.